examples/batched-solver/batched-solver.cpp

```diff
@@ -108,46 +147,205 @@ int main(int argc, char *argv[])
+    auto batch_vec_size =
+        gko::batch_dim<>(num_systems, gko::dim<2>(num_rows, 1));
+    // @sect3{Use of application-allocated memory}
+    // We can either work on the existing memory allocated in the application,
+    //  or we can copy it for the linear solve.
+    //  Note: it is not possible to use data through a const pointer directly.
+    //  Because our pointers are not const, we can just 'wrap' the given
+    //  pointers into Ginkgo Array views so that we can create a Ginkgo matrix
+    //  out of them.
+    // Ginkgo expects the nonzero values for all the small matrices to be
+    //  allocated contiguously, one matrix after the other.
+    auto vals_view = gko::Array<value_type>::view(
+        exec, num_systems * appl_sys.nnz, appl_sys.all_values);
+    auto rowptrs_view =
+        gko::Array<index_type>::view(exec, num_rows + 1, appl_sys.row_ptrs);
+    auto colidxs_view =
+        gko::Array<index_type>::view(exec, appl_sys.nnz, appl_sys.col_idxs);
+    auto A = gko::share(mtx_type::create(exec, batch_mat_size, vals_view,
+                                         colidxs_view, rowptrs_view));
+    // @sect3{Batch stride}
+    // batch_stride object specifies the access stride within the individual
+    //  matrices (vectors) in the batch. In this case, we specify a stride of 1
+    //  as the common value for all the matrices.
+    auto batch_vec_stride = gko::batch_stride(num_systems, 1);
+    // Create RHS, again reusing application allocation
+    auto b_view = gko::Array<value_type>::view(exec, num_systems * num_rows,
+                                               appl_sys.all_rhs);
+    auto b = vec_type::create(exec, batch_vec_size, b_view, batch_vec_stride);
+    // Create initial guess as 0 and copy to device
+    auto x = vec_type::create(exec);
+    auto host_x =
+        vec_type::create(exec->get_master(), batch_vec_size, batch_vec_stride);
+    for (size_type isys = 0; isys < num_systems; isys++) {
+        for (int irow = 0; irow < num_rows; irow++) {
+            host_x->at(isys, irow, 0) = gko::zero<value_type>();
+        }
     }
-    auto x = vec::create(exec);
-    auto b = vec::create(exec);
     x->copy_from(host_x.get());
-    b->copy_from(host_x.get());
 
-    // Create batched matrices from the input matrix and generated vectors.
-    auto bA = share(bmtx::create(exec, nbatch, A.get()));
-    auto bb = bvec::create(exec, nbatch, b.get());
-    auto bx = bvec::create(exec, nbatch, x.get());
-    const RealValueType reduction_factor{1e-7};
+    const real_type reduction_factor{1e-6};
+
+    // @sect3{Create the batch solver factory}
+    // Create a batched solver factory with relevant parameters.
+    auto solver_gen =
+        solver_type::build()
+            .with_max_iterations(500)
+            .with_residual_tol(reduction_factor)
+            .with_tolerance_type(gko::stop::batch::ToleranceType::relative)
+            .with_preconditioner(gko::preconditioner::batch::type::jacobi)
```

Comment:
I am not convinced with it.
At least for preconditioners, it gives different usage from usual ginkgo usage.
It also gives several configs from one of any preconditioner.
the issue also applies tolerance

Reply:
It also changes how we handle the memory of preconditioners.
in a current way, you need to handle different generate methods in if-else and tons of pointer (void maybe).
Jacobi may not introduce too much different performance impact on generating in apply, but it should give the difference in ILU or some not light-way preconditioner

Reply:
The idea here is that we'll only have a few (maybe 2 or 3) preconditioners for batched solvers, and they will be tightly integrated with the solvers. So I don't yet know what we can put in a "core" module for batched preconditioners. So there are no user-facing classes for them, and it makes sense to me to just provide them as factory parameters. We are open to other options though, and the interface can be changed.

Reply:
for example, jacobi contains max_blocks, parilu contains sweeps, or ilut contains threshold, etc...
you will need to add all configurations from possible preconditioners into all solvers.
The preconditioner is not only a preconditioner but also a linop. If it is a linop, it always needs core even if the structure is close to empty.

Reply:
Yes, that may well turn out to be an issue once we want to integrate batch Par ILU or something. We'll need to think of an architecture for that - I'm sure something is possible. I will definitely keep this mind when thinking about the final code structure for this.